plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -190,6 +203,78 @@ public TableSnapshot loadSnapshot(ConnectorSession session, SchemaTableName tabl
         return snapshot;
     }
 
+    private TableSnapshot loadSnapshotForTimeTravel(TrinoFileSystem fileSystem, SchemaTableName table, String tableLocation, long endVersion)
+            throws IOException
+    {
+        try {
+            Optional<LastCheckpoint> lastCheckpoint = findCheckpoint(fileSystem, tableLocation, endVersion);
+            return TableSnapshot.load(
+                    table,
+                    lastCheckpoint,
+                    fileSystem,
+                    tableLocation,
+                    parquetReaderOptions,
+                    checkpointRowStatisticsWritingEnabled,
+                    domainCompactionThreshold,
+                    Optional.of(endVersion));
+        }
+        catch (UncheckedExecutionException e) {
+            throwIfUnchecked(e.getCause());
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Optional<LastCheckpoint> findCheckpoint(TrinoFileSystem fileSystem, String tableLocation, Long endVersion)
+    {
+        Optional<LastCheckpoint> lastCheckpoint = readLastCheckpoint(fileSystem, tableLocation);
+        if (lastCheckpoint.isPresent() && lastCheckpoint.get().version() <= endVersion) {
+            return lastCheckpoint;
+        }
+
+        // TODO Make this logic efficient
+        Optional<LastCheckpoint> latestCheckpoint = Optional.empty();
+        Location transactionDirectory = Location.of(getTransactionLogDir(tableLocation));
+        try {
+            FileIterator files = fileSystem.listFiles(transactionDirectory);
+            while (files.hasNext()) {
+                FileEntry file = files.next();
+                Optional<LastCheckpoint> checkpoint = extractCheckpointVersion(file);
+                if (checkpoint.isEmpty()) {
+                    continue;
+                }
+
+                long version = checkpoint.get().version();
+                if (version > endVersion || (latestCheckpoint.isPresent() && version < latestCheckpoint.get().version())) {
```

Comment:
We shouldn't look for a checkpoint which is greater than the version we're querying.

Reply:
@electrum does the `io.trino.filesystem.TrinoFileSystem#listFiles` guarantees that the listing of the files is being done in lexicographical order?

Reply:
I had the same opinion, but I don't think the order is ensured. 

For instance, org.apache.hadoop.fs.FileSystem#listFiles method says:
> List the statuses and block locations of the files in the given path. Does not guarantee to return the iterator that traverses statuses of the files in a sorted order.